common/test/TestLogging.cs

```diff
@@ -2,39 +2,115 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
-using System.Diagnostics;
-using System.Globalization;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.Versioning;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
 
 namespace Microsoft.Azure.Devices.E2ETests
 {
+    /// <summary>
+    /// This class logs to multiple providers - Event source, Application insights
+    /// </summary>
     public class TestLogging
     {
-        private readonly static TestLogging s_instance = new TestLogging();
+        public const string Language = "C#";
+        public const string Service = "IotHub";
+        private readonly object initLock = new object();
+
+        // Client to log to application insights
+        private static TelemetryClient _telemetryClient;
+
+        // Unique ID for all tests of a run.
+        private static string _testRunId;
+
+        public TestContext Context { get; set; }
+        public string TargetFramework { get; set; }
 
         private TestLogging()
         {
+            // Thread-safe initialization of the telemetry client and run Id.
+            lock (initLock)
+            {
+                if (_telemetryClient == null)
+                {
+                    string iKey = Environment.GetEnvironmentVariable("E2E_IKEY");
```

Comment:
If you do the check outside of the lock, you can skip the lock all together for runs that don't have an AI Key.

Reply:
I did not choose to do that because every test does not have to Get the EnvVariable. And this way, it happens only once. And after I update all pipelines, this EnvVariable will always be there so the check might not be very useful. This is more to guard against errors till I onboard every pipeline.